packages/react-native-codegen/src/parsers/typescript/components/props.js

```diff
@@ -464,6 +480,56 @@ function getTypeAnnotation(
   }
 }
 
+function findProp(
+  name: string,
+  typeAnnotation: $FlowFixMe,
+  optionalType: boolean,
+) {
+  switch (typeAnnotation.type) {
+    // Check for (T)
+    case 'TSParenthesizedType':
+      return findProp(name, typeAnnotation.typeAnnotation, optionalType);
+
+    // Check for optional type in union e.g. T | null | undefined
+    case 'TSUnionType':
+      return findProp(
+        name,
+        typeAnnotation.types.filter(
+          t => t.type !== 'TSNullKeyword' && t.type !== 'TSUndefinedKeyword',
+        )[0],
+        optionalType ||
+          typeAnnotation.types.some(
+            t => t.type === 'TSNullKeyword' || t.type === 'TSUndefinedKeyword',
+          ),
+      );
+
+    case 'TSTypeReference':
+      // Check against optional type inside `WithDefault`
+      if (typeAnnotation.typeName.name === 'WithDefault' && optionalType) {
+        throw new Error(
+          'WithDefault<> is optional and does not need to be marked as optional. Please remove the union of undefined and/or null',
+        );
+      }
+      // Remove unwanted types
+      if (
+        typeAnnotation.typeName.name === 'DirectEventHandler' ||
+        typeAnnotation.typeName.name === 'BubblingEventHandler'
```

Comment:
I wonder whether it makes sense to define some utilities like `isEvent` or `isStyle` and reuse them.
I already saw this `||` multiple times now...

Reply:
I closed the previous one since these two are duplicated. So when codegen sees a component spec file, it assumes interface members are either events or component properties. But the original code handle interface members in such way:

- find all members that are qualified as events, translate them to event schema.
- find all members that are qualified as properties, translate them to property schema.

But member types could be wrapped in a complex structure, like `(T)` and `T | undefined | null` etc. So you could see in the original code, the event part is searching for `DirectEventHandler` and `BubblingEventHandler`, and the property part is ignoring `DirectEventHandler` and `BubblingEventHandler`.

When I refactor the code, I don't want to rewrite the idea too much (actually I think it is better to do that but I think making refactoring in multiple PRs is safer and making more sense). Now you see in `findEvent` and `findProp`, there are `name === 'D' || name === 'B'` and `name !== 'D' && name !== 'B'`. So in the previous comment there is also no "positive predicate".

I will still making more component PRs, I think in a proper timing I will refactor the framework to ensure that, if a member is recognized as an event, then it will not be processed again in the property part, so that many code could be deleted. But such change will be messy, I don't want to mix it with fixing bugs.